### PR TITLE
Add default shortcuts for moving cells

### PR DESCRIPTION
**Pull request by KrishnaKumarHariprasannan**
## References
Fixes https://github.com/jupyterlab/jupyterlab/issues/7447

## Code changes
Added shortcuts to control moving notebook cells up/down

## User-facing changes
Users can now use `Ctrl Shift Up/Down` to move the cell in focus by one cell in either direction

## Backwards-incompatible changes
None

### CODE REVIEW

**Comment by mlucool** [comment]
CC @jasongrout
Any thoughts about this?

**Comment by jasongrout** [comment]
> CC @jasongrout
> Any thoughts about this?

Sorry, at the moment I'm maxed out thinking about JupyterCon right now.

**Comment by mlucool** [comment]
Ah yes I forgot that was starting - thanks for helping put that on!

**Comment by andrewfulton9** [comment]
I tested this PR and it works for me. I can use CTRL+SHFT+[up or down] and it flips the selected cell with the the cell above or below it. I can also select multiple cells and shift all at once with the cell above or below them. Do we want backwards compatibility with the classic notebook? The classic notebook can move cells, but doesn't have the hot keys defined for the action as far as I can tell. cc: @echarles, @goanpeca

**Comment by krassowski** [comment]
I can confirm that these defaults are reasonable on Ubuntu/Gnome and would be a welcome feature for me.

Would it be good for 3.1? The current branch target is 2.2.x rather than master.

**Comment by jasongrout** [comment]
For things like this, we usually gather data about what is done in other editors to see how consistent this would be with other applications.

Ctrl-shift up/down does:
* Visual Studio Code: Nothing
* Sublime text: makes a new multi-cursor one line above/below the current cursor
* Emacs: Extens current highlight by paragraph blocks (natural extension of shift up/down extending the highlight by lines)
* vim: seems to just be the same as up/down, i.e., moving the cursor, in both edit and command mode
* browser text box: nothing
* codemirror: highlights to the start/end of the editor

**Comment by jasongrout** [comment]
As for what does move a line up/down:

* VS Code: Alt up/down
* Sublime: Ctrl Cmd up/down (on macOS)
* Eclipse: Alt up/down
* Emacs: no built-in (C-x C-t transposes), but wiki provides an implementation: https://www.emacswiki.org/emacs/MoveLine
* XCode: Cmd Option [ or ]
* VIM: No builtin shortcut (can use `:m +1`, etc.). https://vim.fandom.com/wiki/Moving_lines_up_or_down suggests mapping Alt j/k
* Apparently Notepad++ uses Ctrl Shift up/down

**Comment by KrishnaKumarHariprasannan** [comment]
Thanks for all that useful info @jasongrout ! To add to the list above:
* PyCharm:  Ctrl Shift Up/Down (from https://github.com/jupyterlab/jupyterlab/issues/7447)
* Atom: Ctrl Up/Down

Popular choices:
Alt Up/Down -> VS Code, Eclipse, Vim (approx. equivalent)
Ctrl Shift Up/Down -> PyCharm, Notepad++

I'll also note that the above choices do not have any conflict with the existing shortcuts and that we use [Shift Up/Down](https://github.com/jupyterlab/jupyterlab/blob/97c8b827561fcec8441be5948a3ec12a20d37cd9/packages/notebook-extension/schema/tracker.json#L81) to extend cell selection. It's also worth thinking about what would be the shortcut for multi-line cursor (when this support arrives) - notebook uses Alt click/drag (https://github.com/jupyter/notebook/issues/4761) - as this seems related to the choice here.

Do we have a preference for either of the above or other outstanding option for moving cells?

> Would it be good for 3.1? The current branch target is 2.2.x rather than master.

Sure, I'll update the target!

**Comment by jasongrout** [comment]
> Do we have a preference for either of the above or other outstanding option for moving cells?

I think this would be a great question for the [weekly dev meeting](https://github.com/jupyterlab/jupyterlab#weekly-dev-meeting)

**Comment by jasongrout** [comment]
In the dev meeting today, we decided to make a decision on the shortcut to use in two weeks at the dev meeting.

* There are now a lot more people aware and trying things out to form an opinion.
* This gives the accessibility group (who is very interested in keyboard shortcuts) a chance to weigh in (they have their meeting after our weekly meeting next week).

**Comment by vidartf** [comment]
Notes from a Windows / Firefox user:
- Trying Ctrl + Shift + Up/Down on a few web interfaces:
  - HackMD: Moves the current line up/down
  - Vanilla Textarea: Selects text to line above, i.e. same as Shift + Up/Down
- Trying Alt + Up/Down:
  - Vanilla Textarea + HackMD: Holding Alt changes cursor to a cross-hair, and the mouse selection mode then changes to "block select":
  image
     Alt + Up/Down does not do anything, i.e. cursor does not move.

To me this would tip in favor of Ctrl+Shift vs Alt.

**Comment by isabela-pf** [comment]
This is a great PR! Thanks for bringing this up at the weekly dev meeting and thanks for your persistence.

I have this on the agenda for next week's JupyterLab accessibility meeting to check if there is some default shortcut preferred by those making use of accessibility features or any concerns about overlapping with default keyboard commands for people using assistive technology. Just a heads up for @manfromjupyter since they have the most expertise of the people that have been working on accessibility in JupyterLab.

Thank you again;  I will get back to you with an answer as soon as I can!

**Comment by manfromjupyter** [comment]
@isabela-pf Thank you for the tag :)

@KrishnaKumarHariprasannan So I did some testing and can confirm it works for non screen readers and screen readers. One thing to note for screen readers is it may appear they can't do it when trying `CTRL+SHIFT+DOWN ARROW/UP ARROW` but it's because both JAWS and NVDA are blocking the command because these naturally negate most keyboard commands sent by the web browser and this happens to be one of them it seems. For JAWS you have to press `INSERT+Z` and for NVDA you have to press `NVDA+SHIFT+SPACE` first but most users needing screen readers will already know this and if you add a label in there to use CTRL+SHIFT+UP and DOWN ARROWs, they will know to try toggling off what is known as the "Virtual Cursor" in JAWS and "browsing mode" in NVDA. So can definitely say it works, just with one annoying intermediate step for the visually impaired/ambulatory.

One accessibility piece that will get fixed but think should be noted is when exiting the cell with "ESC" the jp-Notebook element it lands on has no readable element or label so will return silence to the screen reader and when doing a read from current place it does a vomit of the entire notebook (because it's the outer element) but this should be fixed when #9622 gets merged.

**Side Note:** would be nice to expose this keyboard shortcut to said users in an aria-label or add exposed buttons instead/also but I know it will get super convoluted/be a long conversation as this is all on the outer element, there are dozens of shortcuts that probably also need exposing, and there is no telling how many screen reader users want to have this read to them every time they exit a cell so will say beyond the scope of this pull request/ticket.

As a happy median, would it be possible or also beyond the scope of the ticket to return a success message or visually hidden one with a role="alert" on it saying "Notebook cells shifted down/up successfully." or something like that after registering a successful move? Not sure where all Jupyter alerts are supposed to sprout from but if visually hidden, could literally be anywhere and it will interrupt the user's verbal feedback to read them that alert. This would help one of our accessibility requirements.

**Comment by KrishnaKumarHariprasannan** [comment]
> As a happy median, would it be possible or also beyond the scope of the ticket to return a success message or visually hidden one with a role="alert" on it saying "Notebook cells shifted down/up successfully." or something like that after registering a successful move?

I can check this - please do point me to an existing alert that demonstrates this behaviour (if any).

**Comment by manfromjupyter** [comment]
@KrishnaKumarHariprasannan Thank you. I did some poking around trying to flush one of these out and even asked some other JLab users but no luck. I think all of the non-notebook code errors appear in modals/pop ups. Unless someone else knows of where an inlined error appears or until the Accessibility Task Force adds these alerts to other sections, I think there's nothing to grab inspiration/best practices from. 

I don't know Typescript, but if you're still willing, would just need a `<div class="sr-only" role="alert">Notebook cell shifted down successfully.</div>` or `up successfully` at the bottom or top of the `#jp-Notebook` region. The "sr-only" is a class already added to the product that makes the container hidden from all visual users so will just be for screen readers. And then just overwrite this same element with the same or "up successfully" message if they do it again so can add an `id` to it also. As long as the alert doesn't get deleted until the user refreshes the page, it'll be perfect.

**Comment by jasongrout** [comment]
> I don't know Typescript, but if you're still willing, would just need a `<div class="sr-only" role="alert">Notebook cell shifted down successfully.</div>` or `up successfully` at the bottom or top inside the `#jp-Notebook` container. The "sr-only" is a class already added to the product that makes the container hidden from all visual users so will just be for screen readers. And then just overwrite this same element with the same or "up successfully" message if they do it again so can add an `id` to it also. As long as the alert doesn't get deleted until the user refreshes the page, it'll be perfect.

It sounds like we should build a notification api around this so that people don't have to deal with updating the DOM, the special classes we use, etc.

**Comment by manfromjupyter** [comment]
@jasongrout, that could be pretty handy.

To prevent this notification API proposal piece from being a blocker over him releasing what he has, would it be okay if he made this one change directly in the DOM and then in the feature ticket for this notification API, we mention to pull this out of DOM and convert his JS firing to instead fire (or whatever) in that area (if @KrishnaKumarHariprasannan is up to it of course).

**Comment by KrishnaKumarHariprasannan** [comment]
Hi @manfromjupyter, @jasongrout  ! Sorry about the long pause here - I've implemented an initial version of https://github.com/jupyterlab/jupyterlab/pull/9031#issuecomment-773541469 - please do check and add any feedback. Also, do we want to add similar messages to other notebook actions (Cut Cell, Copy Cell, etc.) ?

@manfromjupyter - I don't have any experience with screen readers, can you please test these changes with one and let me know if you'd like to make any changes?

**Comment by manfromjupyter** [comment]
@KrishnaKumarHariprasannan Binder is still blocking requests from AWS VM/boxes so I can't do it now. I would be happy to though just might take a few hours to get back to you.

**Comment by manfromjupyter** [comment]
@KrishnaKumarHariprasannan So went ahead and tested it and currently when you go from inside the notebook cell to selecting the entire cell, it reads to the user "notebook contents" not "cell contents" which I think might be preferred. Would be even more useful to the user if we had the cells labeled somehow, perhaps by number after a cell has been run to say ex. "cell 5's content". n addition, when moving it merely says "selected" (and that's after they go to a cell, esc it, INSERT+Z to allow custom hotkeys, then CTRL+SHIFT+Up arrow.

To answer your question earlier, no I do not think the editor buttons require alerts because the whole point of alerts is to let the user know something they did not really expect. Could mean a success or error message or a notice, etc. When the user is making an action they expect it to do what it was programmed to do (according to WCAG at least). So hitting the scissors doesn't need to say "Cell has been cut". They should assume it's cut. If it was presumed that a notebook saves or runs but it takes minutes to run or save and users were having problems with knowning when it's done, then yes, those might need an alert, but as is which how fast it is and expected to be, they are not needed. For this because there is no exposed UI for the the arrows and this isn't a normal action, I think an alert is important.

Screenshots: With cells that have been run verse not.
what is says after move and while focused on entire cell
what is says after move and while focused on entire cell after run

## Synopsis of What I think left should be done
1. Add aria-label to each cell labeled "cell content". Bonus points if you tell them what kind of cell (markdown, code, or raw) they are on so they know "Markdown Cell Content" vs "Code Cell Content". In addition, also include the command to use do this action. I believe the ideal notebook cell should say "code cell content use arrow key to move to other cells and ctrl shift arrow keys to swap cells". Lot of words, I know, but it real screenreader users after the first time maybe will not be reading it in full before continuing, they will need to hear merely the first syllable to know what to do. "Mar-" tells them in a millisecond this is a markdown cell and if they are familiar with the notebook, they know where to move it or where the rest of the cells are in relation. Numbers in addition would be ideal but I know that's a bigger discussion probably for @jasongrout. Bonus Bonus points if you also say "Code Cell Content with Output" as the prefix. This will give them even more information to go on (because when you move a cell you also move the output). Helps with other things too not defined in the ticket.
2. Add role="alert" for a piece of screenreader only text in a container that gets replaced every time in the event they do it multiple times in a row that tells them "Cell content successfully swapped with cell above" (or below).

If we can do these things, I'd say you've not only made this feature 100% accessible but you've provided much needed context for screenreaders and would have successfully give them the power to navigate from cell to cell that otherwise does not audibly exist out of the box. Sounds big but if you're up to it, it'll make a real difference. As per the user needing to type INSERT+Z to then do your CTRL+SHIFT+Up arrow, that is not a big deal; upon hearing what shortcut they need to use, most will know they have to enable custom hotkeys temporarily. Anyway, let me know if you need help testing any of these things. Happy to test it for you so it's not a blocker.

File: packages/notebook-extension/src/index.ts [hunk 1]
```diff
@@ -1730,7 +1730,11 @@ function addCommands(
       const current = getCurrent(args);
 
       if (current) {
-        return NotebookActions.moveUp(current.content);
+        NotebookActions.moveUp(current.content);
+        Private.raiseSilentNotification(
+          'Notebook cell shifted up successfully',
```
Comment (krassowski) [inline packages/notebook-extension/src/index.ts:1735]: Needs to be translated:

```suggestion
          trans.__('Notebook cell shifted up successfully'),
```

File: packages/notebook-extension/src/index.ts [hunk 2]
```diff
@@ -1741,7 +1745,11 @@ function addCommands(
       const current = getCurrent(args);
 
       if (current) {
-        return NotebookActions.moveDown(current.content);
+        NotebookActions.moveDown(current.content);
+        Private.raiseSilentNotification(
+          'Notebook cell shifted down successfully',
```
Comment (krassowski) [inline packages/notebook-extension/src/index.ts:1750]: Needs to be translated as well.

**Comment by krassowski** [comment]
Sorry for long silence @KrishnaKumarHariprasannan. I added two minor comments on translations; if you could address those then I am happy to do a final test and approve this one. We can move the discussion on other usability points to a dedicated issue, because while it is great we would ideally address those in a separate pull request.

**Comment by KrishnaKumarHariprasannan** [comment]
Thanks for the review and the suggestion @krassowski ! I've addressed those - the UI test errors seems unrelated (and will likely be fixed with https://github.com/jupyterlab/jupyterlab/issues/10342).

Also, I've moved other enhancements to https://github.com/jupyterlab/jupyterlab/issues/10347 and will address those separately.

**Review by krassowski — APPROVED** [review]
User-experience is good. Code looks good. Introducing `raiseSilentNotification` in notebook-extension might be sub-optimal - I think it would ideally live elsewhere, possibly in a dedicated `@jupyterlab/accessibility` package, but given that this is a proof of concept and it might be better to have it somewhere to begin with, I do not object. Let's see what others think.

**Comment by jweill-aws** [comment]
@krassowski approved this PR over a year ago and, despite the lack of activity since then, this is still ready to merge (no conflicts). Is this safe to merge in?

**Comment by jtpio** [comment]
Let's do a rebase first to make sure CI runs on the latest?

image

**Comment by KrishnaKumarHariprasannan** [comment]
> Let's do a rebase first to make sure CI runs on the latest?

Good idea @jtpio ! Let me get that done.

**Comment by KrishnaKumarHariprasannan** [comment]
There are couple of failures from visual regression test:

```
2 failed
    [jupyterlab] › test/jupyterlab/menus.test.ts:30:5 › General Tests › Open menu item Edit ========
    [jupyterlab] › test/jupyterlab/notebook-create.test.ts:53:5 › Notebook Create › Open menu item Edit
```

The failures seem unrelated to this change - `Expected an image 350px by 657px, received 358px by 657px.`.

To be sure, I wanted to run these tests locally but I'm running into some blockers along the way. Particualrly, I need help accessing the (expected) artifacts of these test - can someone please guide me here?

**Comment by jtpio** [comment]
> Particualrly, I need help accessing the (expected) artifacts of these test - can someone please guide me here?

You can access the ones from the CI run here: https://github.com/jupyterlab/jupyterlab/actions/runs/2890608374

image

After looking at the report, the failure seems relevant because the new shortcuts are displayed in the menu items:

image

**Comment by jtpio** [comment]
Let's use the bot to update the reference snapshots.

**Comment by jtpio** [comment]
bot please update snapshots

**Comment by jtpio** [comment]
Looks like the bot pushed some unrelated changes:

image

Will run the script again so to see if they can be reverted.

**Comment by jtpio** [comment]
bot please update documentation snapshots

**Comment by fcollonval** [comment]
> Will run the script again so to see if they can be reverted.

it is best to update the commit manually - just did it - I will cancel the update run.

**Review by fcollonval — APPROVED** [review]
Thanks @KrishnaKumarHariprasannan for the patience

**Comment by KrishnaKumarHariprasannan** [comment]
Thanks @jtpio, @fcollonval, @jweill-aws ! :)

**Pull request merged**